[Voice to Content] Launch edit post with AI content

zwarm opened the pull request:
Closes https://github.com/Automattic/wordpress-mobile/issues/78

This PR launches the editPost view after successful recording of audio. 

This branch build upon issue/v2c-show-elapsed-time
> [!IMPORTANT]
> **Merge Instructions**
> ~- Make sure #20983 & #20984 have been merged to trunk~
> ~- Remove the Not Ready for Merge Label~
> - Merge as normal

> [!NOTE]
> The following have not yet been implemented:
>  - Orientation changes
>  - Overall polishing of the UI
> - Unit tests

-----
## To Test:
- Install and launch the app
- Login with an account that has access to AI credits (Any a8c P2 will have this - other free sites have 20 requests)
- Navigate to Me > Debug Settings and enable the voice_to_content flag (restart the app)
- Navigate to My Site and tap the FAB
- Select the Post with Audio option
- Start the recording session by tapping on the mic icon
- Record a blog post - a few words, maybe include a list, try it out
- Tap the done icon when finished
- ✅ Verify the processing view is shown
- ✅ Verify the Edit Post view is shown with your recorded content

-----

## Regression Notes

1. Potential unintended areas of impact
The recorded content is not shown

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
N/A

-----

## PR Submission Checklist:
- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist (strike-out the not-applying and unnecessary ones): N/A

zwarm committed:
Add constant for EXTRA_CONTENT_FROM_AI

zwarm committed:
Add new method to launch editPost with content from AI

zwarm committed:
Add state key for voice to content and update extra name

zwarm committed:
Add updateContent as the passthrough to the container for voice to content

zwarm committed:
Implement updateContent() for voice to content - GB only

zwarm committed:
Update to reflect refactored intent extra name

zwarm committed:
Implement updateVoiceContentIfNeeded to updateContent if this is a voice to content request

zwarm committed:
Holds VoiceToContent actions and events

zwarm committed:
Observe actionEvents from the viewModel, implement launchEditPost

zwarm committed:
Create actionEvents liveData and post launchEditPost on success of processing v2c

zwarm committed:
Refactor: combine liveData into actionEvents, clean up log lines, and unneeded comments

zwarm committed:
Address detekt issues

zwarm committed:
Address checkstlye be removing empty line

dangermattic commented:
<!--
  0 Errors
  2 Warnings: Class <code>VoiceToContentActi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>VoiceToContentActionEvent</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano reviewed (commented):
- When I add a self-hosted site without logging in with a WP.com account, I see this screen. You can address this in another PR, but let's not forget to hide the button for self-hosted sites.
<img src="https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/c30fd24c-990b-4d21-8444-41f089f0e9d4" width=300>

- I am supposed to see "Edit Post" after a successful recording session, right? I can't see it.

[record-issue.webm](https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/51904ab3-5f31-4938-bf81-1e7a93198096)

zwarm committed:
Remove errant dismiss call

zwarm commented:
@irfano - The edit post is fixed - I caught that and fixed it in a future PR, I added it in [2758f22](https://github.com/wordpress-mobile/WordPress-Android/pull/20992/commits/2758f22519d8a0bac2bcdf4c842a9d059256c0be). 
Thanks for the heads up on the self-hosted site. As long as the site has AI credits it can work, but I will double check - this error could be because of a jetpaack-ai-assistant API failure, in which case this is the correct error message (but not formatted pretty :)

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:
```diff
@@ -717,6 +718,7 @@ class EditPostActivity : LocaleAwareActivity(), EditorFragmentActivity, EditorIm
                 }
 
             isNewPost = state.getBoolean(EditPostActivityConstants.STATE_KEY_IS_NEW_POST, false)
+            isVoiceContentSet = state.getBoolean(EditPostActivityConstants.STATE_KEY_IS_NEW_POST, false)
```
`isVoiceContentSet` uses the wrong key here, it should use `STATE_KEY_IS_VOICE_CONTENT_SET`.

zwarm replied:
This is a great find, thank you. This is what happens when I am rushing to get most of the work into the repo before I switch teams. :(

zwarm replied:
Updated in [82a6ca6](https://github.com/wordpress-mobile/WordPress-Android/pull/20992/commits/82a6ca668a10a96036538c15cabfba43d8769a30)

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:
```diff
@@ -3526,6 +3529,20 @@ class EditPostActivity : LocaleAwareActivity(), EditorFragmentActivity, EditorIm
+    private fun updateVoiceContentIfNeeded() {
+        // Check if voice content exists and this is a new post for a Gutenberg editor fragment
+        val content = intent.getStringExtra(EditPostActivityConstants.EXTRA_VOICE_CONTENT)
+        if (isNewPost && content != null && !isVoiceContentSet) {
```
Do we need to track this `isVoiceContentSet` with a boolean? Is there a case this `updateVoiceContentIfNeeded()` is being called multiple times for a single `EditPostActivity` instance?

zwarm replied:
@irfano To prevent the content from reloading before the post is saved - could happen on return from background.

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/voicetocontent/VoiceToContentDialogFragment.kt:
```diff
@@ -78,17 +83,16 @@ class VoiceToContentDialogFragment : BottomSheetDialogFragment() {
 
         // Observe the ViewModel to update the cancelable state of closing on outside touch
         viewModel.isCancelableOutsideTouch.observe(this) { cancelable ->
-            Log.i(javaClass.simpleName, "***=> disable outside touch")
```
Good to remove this log. I also thought it was unnecessary. 👍🏻

zwarm replied:
:) My final PR cleans up other log lines and debug comments/helpers, too!

irfano replied:
Thank you!

zwarm committed:
Fix using the wrong state key

irfano reviewed (approved):
Thank you for your hard work on this project. LGTM! 👍🏻
